online-shop/src/app/edit-product-form/edit-product-form.component.ts

```diff
@@ -0,0 +1,62 @@
+import { Location } from '@angular/common';
+import { Component, OnInit } from '@angular/core';
+import { FormBuilder, Validators } from '@angular/forms';
+import { ActivatedRoute } from '@angular/router';
+import { Product } from '../data/product';
+import { ProductService } from '../services/product.service';
+
+@Component({
+  selector: 'app-edit-product-form',
+  templateUrl: './edit-product-form.component.html',
+  styleUrls: ['./edit-product-form.component.scss'],
+})
+export class EditProductFormComponent implements OnInit {
+  productId = -1;
+  detailsForm = this.fb.nonNullable.group({
+    name: ['', Validators.required],
+    category: ['', Validators.required],
+    price: [0, [Validators.required, Validators.min(1)]],
+    image: ['', Validators.required],
+    description: ['', Validators.required],
+  });
+
+  constructor(
+    private fb: FormBuilder,
+    private route: ActivatedRoute,
+    private location: Location,
+    private productService: ProductService
+  ) {}
+
+  ngOnInit(): void {
+    this.productId = Number(this.route.snapshot.paramMap.get('id'));
+
+    this.productService
+      .getProduct(this.productId)
```

Comment:
**suggestion:** take care to avoid memory leaks in your code by unsubscribing from observables. https://www.thisdot.co/blog/best-practices-for-managing-rxjs-subscriptions/